src/main/java/com/openshift/jenkins/plugins/pipeline/OcWatch.java

```diff
@@ -95,132 +93,90 @@ public boolean takesImplicitBlockArgument() {
-                                if (o instanceof Boolean && (Boolean)o) {
-                                    listener.getLogger().println("\nwatch closure returned true; terminating watch");
-                                    proc.kill();
-                                    break master;
-                                }
-                                listener.getLogger().println("watch closure returned " + o);
-                                jsonRetries = 0;
-
-                            } catch ( InterruptedException tie ) { // timeout{} block interrupted us
-                                listener.getLogger().println("\nwatch closure interrupted (timeout?)");
-                                getContext().onFailure(tie);
-                                return null;
-                            } catch (Throwable t) {
-                                if (t instanceof groovy.json.JsonException) {
-                                    // we've seen instances where if the watch closer susequently calls oc and it processes output 
-                                    // before oc has finished updating it we can get json formatting 
-                                    // exceptions processing the output ... we can retry here
-                                    if (jsonRetries < 5) {
+                                for (; ; ) {
+                                    listener.getLogger().println("Running watch closure body");
+                                    try {
+                                        // Run body and get result
+                                        Object o = getContext().newBodyInvoker().start().get();
+
+                                        // If the watch body returns a Boolean and it is true, time to exit
+                                        if (o instanceof Boolean && (Boolean) o) {
+                                            watchSuccess[0] = true;
+                                            listener.getLogger().println("\nwatch closure returned true; terminating watch");
+                                            return true; // interrupt `oc`
+                                        }
+                                        listener.getLogger().println("watch closure returned " + o);
+                                        return false; // don't interrupt `oc`
+                                    } catch (groovy.json.JsonException ex) {
+                                        // FIXME: We've seen instances where if the watch closer subsequently calls oc and it processes output
+                                        // before oc has finished updating it we can get json formatting
+                                        // exceptions processing the output ... we can retry here
+                                        if (jsonRetries[0]++ >= 5) {
+                                            getContext().onFailure(ex);
```

Comment:
As our exchanges were occurring at the same time, I had landed with being OK with the catch all getContext().onFailure() call in the final try / catch (I had missed that in my initial review).

We don't need to worry about it with this PR ... I'll see about addressing in a follow up.

Reply:
Unless there are more changes coming as part of the review process ....

Reply:
OK, removed those `getContext().onFailure()` calls, and also added the catch all getContext().onFailure() call to OcAction.